sympy/physics/units/dimensions.py

```diff
@@ -437,18 +438,32 @@ def _get_dimensional_dependencies_for_name(self, name):
             raise TypeError("Only equivalent dimensions can be added or subtracted.")
 
         if name.is_Pow:
-            dim = get_for_name(name.base)
-            return {k: v*name.exp for (k, v) in dim.items()}
+            dim_base = get_for_name(name.base)
+            dim_exp = get_for_name(name.exp)
+            if dim_exp != {}:
+                raise TypeError("The exponent for the power operator must be dimensionless.")
+            return {k: v*name.exp for (k, v) in dim_base.items()}
 
         if name.is_Function:
             args = (Dimension._from_dimensional_dependencies(
                 get_for_name(arg)) for arg in name.args)
             result = name.func(*args)
 
+            dicts = [get_for_name(i) for i in name.args]
+
             if isinstance(result, Dimension):
                 return self.get_dimensional_dependencies(result)
             elif result.func == name.func:
-                return {}
+                if name.func in (sin, cos, tan, cot, sec, csc):
```

Comment:
this should be `isinstance(name, (sin, cos, tan, cot, sec, csc))`, because using `name.func` is not aware of subclassing and may change behaviour in the future. There are some discussions somewhere on why it's always better to use `isinstance`.

Reply:
Makes sense, I've updated it to reflect this change.